src/dag-node.js

```diff
@@ -1,281 +1,79 @@
-    })
-
-    Object.defineProperty(this, 'size', {
-      get () {
-        return size
-      },
-      set: immutableError
-    })
-
-    Object.defineProperty(this, 'multihash', {
-      get () {
-        return multihash
-      },
-      set: immutableError
-    })
-
-    Object.defineProperty(this, 'json', {
-      get () {
-        return json
-      },
-      set: immutableError
-    })
-
-    this.link = {
-      add: (nameOrLink, nodeOrMultihash, callback) => {
-        if ((nameOrLink.constructor &&
-             nameOrLink.constructor.name === 'DAGLink')) {
-          // It's a link
-          const link = nameOrLink
-          // TODO
-          // 1. add to the set of links
-          // 2. create new node
-        } else if (typeof nameOrLink === 'string') {
-          // It's a name
-          const name = nameOrLink
-          if ((nodeOrMultihash.constructor &&
-             nodeOrMultihash.constructor.name === 'DAGNode')) {
-            // It's a node
-            // TODO
-            // 1. create link from name + node
-            // 2. add to the set of links
-            // 3. create new node
-          } else {
-            // It's a multihash
-            let multihash = nodeOrMultihash
-            if (typeof multihash === 'string') {
-              multihash = mh.fromB58String(multihash)
-            }
-            // TODO
-            // 1. create link from name + multihash
-            // 2. add to the set of links
-            // 3. create new node
-          }
-        }
-        callback(new Error('invalid arguments'))
-      },
-      rm: (nameOrMultihash, callback) => {
-        let links
-        if (typeof nameOrMultihash === 'string') {
-          const name = nameOrMultihash
-          links = this.links.filter((link) => {
-            if (link.name === name) {
-              return false
-            } else {
-              return true
-            }
-          })
-        } else if (Buffer.isBuffer(nameOrMultihash)) {
-          const multihash = nameOrMultihash
-          links = this.links.filter((link) => {
-            if (link.hash.equals(multihash)) {
-              return false
-            } else {
-              return true
-            }
-          })
-        } else {
-          callback(new Error('first arg needs to be a name or multihash'))
-        }
-
-        let dataClone
-
-        if (this.data && this.data.length > 0) {
-          dataClone = new Buffer(this.data.length)
-          this.data.copy(dataClone)
-        }
-
-        DAGNode.create(dataClone, links, callback)
-      }
+    this._data = data || new Buffer(0)
+    this._links = links || []
+    this._serialized = serialized || new Buffer(0) // TODO: default serialized object
+    this._multihash = multihash || new Buffer(0) // TODO: default multihash object
+    this._json = {
+      data: this.data,
+      links: this.links.map((l) => l.json),
+      hash: mh.toB58String(this.multihash),
+      size: this.size
```

Comment:
This makes it so that folks can change the values by doing `node._data = something` while in the previous case that would not happen. Are we happy with that?